Add documentation for SDFormat/Gazebo extension to URDF

azeey opened the pull request:
There is good documentation for the `<gazebo>` tag at https://classic.gazebosim.org/tutorials?tut=ros_urdf&cat=connect_ros, but it is presented in a tutorial format and is missing some concepts and explanations. This is an attempt to document the current behavior and provide an explanation about fixed joint lumping.


[Preview](http://sdformat.org/tutorials?tut=sdformat_urdf_extensions&cat=specification&branch=urdf_sdf_extension&repo=https%3A%2F%2Fgithub.com%2Fazeey%2Fsdf_tutorials)

aaronchongth commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+
+
+
+The `<material>` tag, when used direclty under the `<gazebo>` tag, i.e
```
```suggestion
The `<material>` tag, when used directly under the `<gazebo>` tag, i.e
```

aaronchongth commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+  <tr>
+    <td>laserRetro</td>
+    <td>double</td>
+    <td>intensity value returned by laser sensor.</td>
```
```suggestion
    <td>Intensity value returned by laser sensor.</td>
```

aaronchongth reviewed (changes requested):
Thanks for the write-up! I've found some typos here and there, but overall it looks very informative.

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+    <static>true</static>
+    <plugin name='testPlugin' filename='testFileName'/>
+  </gazebo>
+</robot>
```
how do you feel about using `<include/>` elements in our Markdown files instead of duplicating the example code?

for example:

* https://classic.gazebosim.org/tutorials?tut=drcsim_animate_joints&cat=drcsim#ModelPluginController
* https://github.com/osrf/gazebo_tutorials/blame/0059020f4b86ae8f736104ef8f6c164f8830bd37/drcsim_animate_joints/tutorial.md#L46

I believe sdformat.org supports the same syntax, though I guess the links may have to point to your fork until the pull request is merged

azeey replied:
Thanks for the pointer. I was hoping there was a way to do that, but I didn't see any examples in sdf_tutorials. I've updated all the examples. https://github.com/gazebosim/sdf_tutorials/pull/88/commits/3beee03fc26bf4c00d3002e39380d7672c7351ca

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+    <td><code>gravity</code></td>
+  </tr>
+  <tr>
+    <td>dampingFactor</td>
```
I'd like to see some padding between the columns of this table

<img width="202" alt="Screen Shot 2023-03-13 at 4 21 27 PM" src="https://user-images.githubusercontent.com/3650755/224853667-e5ef7f28-d2d4-4121-a510-46d892ec0c2a.png">

scpeters replied:
it could also be helpful to draw lines between the rows and columns

azeey replied:
Yes, I had the same idea: https://github.com/osrf/sdf_web_app/pull/66

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+  <tr>
+    <td>turnGravityOff</td>
+    <td>bool</td>
+    <td>A value of "true" turns gravity off. Alternatively, <code>gravity</code> (with opposite boolean value) can be used via blob insertion</td>
```
I think this is the only place blob insertion is mentioned in the document, but it is not defined, so we should explain what it means

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+  <tr>
+    <td>maxVel</td>
+    <td>double</td>
+    <td>maximum contact correction velocity truncation term.</td>
```
consider linking to the following gazebo-classic tutorial for the relevant parameters:

https://classic.gazebosim.org/tutorials?tut=physics_params&cat=physics#Contactparameters

azeey replied:
Good idea! Done in https://github.com/gazebosim/sdf_tutorials/pull/88/commits/3beee03fc26bf4c00d3002e39380d7672c7351ca

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,822 @@
+      <mass value='0.12' />
+      <inertia ixx='0.01' ixy='0' ixz='0' iyy='0.01' iyz='0' izz='0.01' />
+    </inertial>
+    <collision>
```
add another collision to show that all collisions get the parameter?

azeey replied:
Done in https://github.com/gazebosim/sdf_tutorials/pull/88/commits/3beee03fc26bf4c00d3002e39380d7672c7351ca

scpeters reviewed (commented):
it could be nice to add examples for inserting lights and sensors to a link, and a force-torque sensor to a joint using the gazebo tags

azeey commented:
I've addressed all the feedback. I've also added more changes regarding `preserveFixedJoint` since my previous understanding of the parameter was incorrect.

aaronchongth reviewed (approved):
Thanks for going into more detail about the various behaviors of fixed joint lumping. LGTM!

azeey commented:
@scpeters The web app has been deployed and the preview is showing the improved table formatting. Mind giving this another look?

scpeters commented on urdf/sdf_extensions.md:
```diff
@@ -0,0 +1,793 @@
+
+**Example:**
+
+<include src='https://github.com/azeey/sdf_tutorials/raw/urdf_sdf_extension/urdf/examples/link_ref_example.urdf' />
```
```suggestion
<include src='https://github.com/azeey/sdf_tutorials/raw/urdf_sdf_extension/urdf/examples/material_example.urdf' />
```

azeey replied:
https://github.com/gazebosim/sdf_tutorials/pull/88/commits/8d1ad1715d8576d47d854e6474d6c2c282efd925

scpeters commented on urdf/examples/preserve_fixed_joint_example.urdf:
```diff
@@ -0,0 +1,36 @@
+    <origin xyz='0 0 1' rpy='0 0 0'/>
+  </joint>
+  <link name='end_effector'>
+    <inertial>
```
`<inertia />` block is missing, which deviates from prior example

azeey replied:
Good catch. Changed in https://github.com/gazebosim/sdf_tutorials/pull/88/commits/8d1ad1715d8576d47d854e6474d6c2c282efd925. The resulting SDFormat now also has the `end_effector_visual`. Not sure why it was missing before.

scpeters reviewed (commented):
I don't see a use of the `material_visual_example.urdf`

azeey commented:
> I don't see a use of the material_visual_example.urdf

Thanks for catching this. Added in https://github.com/gazebosim/sdf_tutorials/pull/88/commits/1fa504492adb10c308d4cd844999239673643374
